subsys/net/ip/l2/ieee802154/ieee802154_security.c

```diff
@@ -108,14 +108,21 @@ bool ieee802154_decrypt_auth(struct ieee802154_security_ctx *sec_ctx,
 	nonce[12] = sec_ctx->level;
 
 	pkt.in_buf = decrypt_payload_len ? frame + auth_payload_len : NULL;
-	pkt.in_len = decrypt_payload_len;
-	pkt.out_buf = frame;
+	/* payload lengths without the length of the tag */
+	pkt.in_len = decrypt_payload_len -
+		     sec_ctx->dec.mode_params.ccm_info.tag_len;
+	/*
+	 * TC does not copy the header, so prevent the header from being
+	 * overwritten.
+	 */
+	pkt.out_buf = frame + auth_payload_len;
```

Comment:
That's seems to be pointing out of the buffer.

If none of the tc or mtls shim can do in-place, you'll need to handle that case here: instead of frame, providing a pointer on a local buffer. Which you'll copy the result back to the frame then.